### PR TITLE
Feature/sup 34 security

### PR DESCRIPTION
**Pull request by IrynaZazyba**
implemented a custom authentication scenario with Spring Security by adding an extra field to the standard login process

### CODE REVIEW

**Review by alesiahlushakova — APPROVED** [review]

File: src/main/java/by/itech/lab/supplier/auth/CustomUserDetailsService.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+@Service("userDetailsService")
+public class CustomUserDetailsService implements UserDetailsService {
+
+    private final UserRepository userRepository;
+    private final CustomerMapper customerMapper;
+
+    public CustomUserDetailsService(UserRepository userRepository,
+                                    CustomerMapper customerMapper) {
+        this.userRepository = userRepository;
+        this.customerMapper=customerMapper;
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/auth/CustomUserDetailsService.java]: spaces?

File: src/main/java/by/itech/lab/supplier/auth/CustomUserDetailsService.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    @Override
+    public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
+        User user = userRepository.findByEmail(email);
+        UserImpl userImpl;
+        if (user == null) {
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/auth/CustomUserDetailsService.java]: Optional...orElseThrow

File: src/main/java/by/itech/lab/supplier/auth/CustomUserDetailsService.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+            customers.add(customerMapper.map(user.getCustomer()));
+            userImpl = new UserImpl(user.getUsername(), user.getPassword(), customers, true,
+                    true, true, user.isActive(), defineAuthority(user));
+        }
+        return userImpl;
+    }
+
+    private List<GrantedAuthority> defineAuthority(User user) {
+        final List<GrantedAuthority> authorities = new ArrayList<>();
+        authorities.add(new SimpleGrantedAuthority(user.getRole().getAuthority()));
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/auth/CustomUserDetailsService.java]: is it possible to use Collections.singletonList() ?

File: src/main/java/by/itech/lab/supplier/auth/CustomUserDetailsService.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    @Override
+    public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
+        User user = userRepository.findByEmail(email);
+        UserImpl userImpl;
+        if (user == null) {
+            throw new UsernameNotFoundException(String.format("Username not found for email=%s", email));
+        } else {
+            List<CustomerDto> customers = new ArrayList<>();
+            customers.add(customerMapper.map(user.getCustomer()));
+            userImpl = new UserImpl(user.getUsername(), user.getPassword(), customers, true,
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/auth/CustomUserDetailsService.java]: if you're hardcoded default values for expiration and locks, maybe you can remove it from the default constructor and set them it in the UserImpl?

File: src/main/java/by/itech/lab/supplier/auth/config/WebSecurityConfig.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+    public WebSecurityConfig(UserDetailsService userDetailsService, AuthenticationSuccessHandler authenticationSuccessHandler) {
+        this.userDetailsService = userDetailsService;
+        this.authenticationSuccessHandler = authenticationSuccessHandler;
+    }
+
+    @Override
+    protected void configure(HttpSecurity http) throws Exception {
+        http.csrf().disable()
+                    .authorizeRequests()
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/auth/config/WebSecurityConfig.java]: if I'll press Ctrl + Alt + L - what gonna happen with these spaces? Is there some better way to do such separation?

File: src/main/java/by/itech/lab/supplier/auth/handler/CustomAuthenticationSuccessHandler.java [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    public CustomAuthenticationSuccessHandler(ObjectMapper mapper) {
+        this.mapper = mapper;
+    }
+
+    @Override
+    public void onAuthenticationSuccess(HttpServletRequest request,
+                                        HttpServletResponse response, Authentication authentication)
+            throws IOException {
+        UserImpl principal = (UserImpl) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
+        response.getWriter().append(mapper.writeValueAsString(principal.getCustomer()));
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/auth/handler/CustomAuthenticationSuccessHandler.java]: image
what about using built-in mechanisms for writing value in the response? is this not possible here?

File: src/main/java/by/itech/lab/supplier/domain/Address.java [hunk 7]
```diff
@@ -1,19 +1,15 @@
 package by.itech.lab.supplier.domain;
 
-import lombok.AllArgsConstructor;
-import lombok.Builder;
-import lombok.Data;
-import lombok.NoArgsConstructor;
+import lombok.*;
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/domain/Address.java]: WILDCARD IMPORTS!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

File: src/main/java/by/itech/lab/supplier/domain/Role.java [hunk 8]
```diff
@@ -1,9 +1,12 @@
...
 
-public enum Role {
-    SYSTEM_ADMIN("System admin"), ADMIN("Admin"), DISPATCHER("Dispatcher"),
-    LOGISTICS_SPECIALIST("Logistics specialist"), DRIVER("Driver"),
-    DIRECTOR("Director");
+import org.springframework.security.core.GrantedAuthority;
+
+public enum Role implements GrantedAuthority {
+
+    ROLE_SYSTEM_ADMIN("System admin"), ROLE_ADMIN("Admin"), ROLE_DISPATCHER("Dispatcher"),
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/domain/Role.java]: one line - one value

File: src/main/java/by/itech/lab/supplier/domain/User.java [hunk 9]
```diff
@@ -1,20 +1,15 @@
 package by.itech.lab.supplier.domain;
 
-import lombok.AllArgsConstructor;
-import lombok.Builder;
-import lombok.Data;
-import lombok.NoArgsConstructor;
-import org.springframework.security.core.GrantedAuthority;
-import org.springframework.security.core.userdetails.UserDetails;
+import lombok.*;
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/domain/User.java]: image
you must be kidding

File: src/main/java/by/itech/lab/supplier/domain/User.java [hunk 10]
```diff
@@ -110,7 +106,7 @@ public void setRole(Role role) {
         this.role = role;
     }
 
-    @ManyToOne
+    @ManyToOne(fetch = FetchType.EAGER)
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/domain/User.java]: do you really need this data always?

File: src/main/java/by/itech/lab/supplier/domain/User.java [hunk 11]
```diff
@@ -120,7 +116,7 @@ public void setAddress(Address address) {
         this.address = address;
     }
 
-    @ManyToOne
+    @ManyToOne (fetch = FetchType.EAGER)
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/domain/User.java]: do you really need this data always?

File: src/main/java/by/itech/lab/supplier/domain/User.java (hunk 10 above)
Comment (svareyko) [reply src/main/java/by/itech/lab/supplier/domain/User.java]: this is additional query in the database when you will load that entity (for each entity)

File: src/main/java/by/itech/lab/supplier/domain/User.java [hunk 12]
```diff
@@ -130,7 +126,7 @@ public void setCustomer(Customer customer) {
         this.customer = customer;
     }
 
-    @ManyToOne
+    @ManyToOne (fetch = FetchType.EAGER)
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/domain/User.java]: do you really need this data always?

File: src/main/java/by/itech/lab/supplier/domain/User.java [hunk 13]
```diff
@@ -140,7 +136,7 @@ public void setWarehouse(Warehouse warehouse) {
         this.warehouse = warehouse;
     }
 
-    @OneToMany(mappedBy = "createdByUsers")
+    @OneToMany(mappedBy = "createdByUsers", fetch = FetchType.EAGER)
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/domain/User.java]: do you really need this data always?

File: src/main/java/by/itech/lab/supplier/domain/User.java [hunk 14]
```diff
@@ -149,7 +145,7 @@ public void setCreatorWayBills(Set<WayBill> creatorWayBills) {
         this.creatorWayBills = creatorWayBills;
     }
 
-    @OneToMany(mappedBy = "updatedByUsers")
+    @OneToMany(mappedBy = "updatedByUsers", fetch = FetchType.EAGER)
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/domain/User.java]: do you really need this data always?

File: src/main/java/by/itech/lab/supplier/domain/User.java [hunk 15]
```diff
@@ -158,7 +154,7 @@ public void setUpdatorWayBills(Set<WayBill> updatorWayBills) {
         this.updatorWayBills = updatorWayBills;
     }
 
-    @OneToMany(mappedBy = "driver")
+    @OneToMany(mappedBy = "driver", fetch = FetchType.EAGER)
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/domain/User.java]: do you really need this data always?

File: src/main/java/by/itech/lab/supplier/domain/User.java [hunk 16]
```diff
@@ -168,7 +164,7 @@ public void setDriverWayBills(Set<WayBill> driverWayBills) {
     }
 
 
-    @OneToMany(mappedBy = "createdByUsers")
+    @OneToMany(mappedBy = "createdByUsers", fetch = FetchType.EAGER)
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/domain/User.java]: do you really need this data always?

File: src/main/java/by/itech/lab/supplier/domain/User.java [hunk 17]
```diff
@@ -177,7 +173,7 @@ public void setCreatorApplications(Set<Application> creatorApplications) {
         this.creatorApplications = creatorApplications;
     }
 
-    @OneToMany(mappedBy = "lastUpdatedByUsers")
+    @OneToMany(mappedBy = "lastUpdatedByUsers", fetch = FetchType.EAGER)
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/domain/User.java]: do you really need this data always?

File: src/main/resources/db/changelog/1.0/2020-10-19-create-system-admin.yaml [hunk 18]
```diff
@@ -0,0 +1,25 @@
+databaseChangeLog:
+  - changeSet:
+      id: IN-CUST-1
+      author: iryna_zazybo
+      changes:
+        - sql:
+            sql: insert into customer (id, name, registration_date, status) values (3, "SystemCompany", "2020-10-19", "active")
```
Comment (svareyko) [inline src/main/resources/db/changelog/1.0/2020-10-19-create-system-admin.yaml]: you can use external SQL file for such insertions

File: src/main/resources/db/changelog/1.0/2020-10-19-create-system-admin.yaml [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+  - changeSet:
+      id: IN-CUST-2
+      author: iryna_zazybo
+      changes:
+        - sql:
+            sql: insert into zone (id, location) values (3, "NY")
```
Comment (svareyko) [inline src/main/resources/db/changelog/1.0/2020-10-19-create-system-admin.yaml]: for all of them

File: src/main/resources/db/changelog/1.0/2020-10-19-change-table-user.yaml [hunk 20]
```diff
@@ -26,4 +26,12 @@ databaseChangeLog:
...
\ No newline at end of file
+                    nullable: false
+  - changeSet:
+      id: CH-U-23
+      author: iryna_zazybo
+      changes:
+        - modifyDataType:
+            columnName: password
+            newDataType: varchar(150)
+            tableName: user
```
Comment (svareyko) [inline src/main/resources/db/changelog/1.0/2020-10-19-change-table-user.yaml]: empty line at the end?

File: src/main/resources/liquibase.properties [hunk 21]
```diff
@@ -0,0 +1,5 @@
+url=jdbc:postgresql://localhost:5432/test
+username=postgres
+password=1111
```
Comment (vareyko-itechart) [inline src/main/resources/liquibase.properties:3]: ooo, high secure system :-)

**Review by svareyko — CHANGES_REQUESTED** [review]

File: pom.xml [hunk 22]
```diff
@@ -75,6 +84,13 @@
                 <groupId>org.springframework.boot</groupId>
                 <artifactId>spring-boot-maven-plugin</artifactId>
             </plugin>
+            <plugin>
+                <groupId>org.liquibase</groupId>
+                <artifactId>liquibase-maven-plugin</artifactId>
+                <configuration>
+                    <propertyFile>src/main/resources/liquibase.properties</propertyFile>
```
Comment (svareyko) [inline pom.xml]: maybe move it to application.yml ?

File: src/main/java/by/itech/lab/supplier/auth/handler/CustomAuthenticationSuccessHandler.java (hunk 6 above)
Comment (IrynaZazyba) [reply src/main/java/by/itech/lab/supplier/auth/handler/CustomAuthenticationSuccessHandler.java]: what do you mean? 
something like this:    .failureHandler((req, resp, ex) -> resp.setStatus(SC_FORBIDDEN))

File: src/main/java/by/itech/lab/supplier/auth/CustomUserDetailsService.java (hunk 4 above)
Comment (IrynaZazyba) [reply src/main/java/by/itech/lab/supplier/auth/CustomUserDetailsService.java]: not sure that understand what you want

File: src/main/java/by/itech/lab/supplier/auth/handler/CustomAuthenticationSuccessHandler.java (hunk 6 above)
Comment (svareyko) [reply src/main/java/by/itech/lab/supplier/auth/handler/CustomAuthenticationSuccessHandler.java]: there is a lot of ways to do this, you need just choose one of them and use, but dont use low level stuff

File: src/main/java/by/itech/lab/supplier/auth/CustomUserDetailsService.java (hunk 4 above)
Comment (svareyko) [reply src/main/java/by/itech/lab/supplier/auth/CustomUserDetailsService.java]: you always pass `boolean enabled, boolean accountNonExpired, boolean credentialsNonExpired` as true, is it correct?

Comment (IrynaZazyba) [reply src/main/java/by/itech/lab/supplier/auth/CustomUserDetailsService.java]: it's possibility add this functionality to user in the future

Comment (svareyko) [reply src/main/java/by/itech/lab/supplier/auth/CustomUserDetailsService.java]: so, when you will add this then you extend, for now, I see that its hardcoded and you do not necessary stuff

> there is nothing more permanent than temporary

File: src/main/resources/db/changelog/1.0/2020-10-05-create-tables.yaml [hunk 23]
```diff
@@ -947,10 +947,4 @@ databaseChangeLog:
...
             initiallyDeferred: false
             referencedColumnNames: id
-            referencedTableName: write_off_act_reason
-
-
-
-
-
-
+            referencedTableName: write_off_act_reason
```
Comment (svareyko) [inline src/main/resources/db/changelog/1.0/2020-10-05-create-tables.yaml]: empty line?

File: src/main/java/by/itech/lab/supplier/auth/config/WebSecurityConfig.java [hunk 24]
```diff
@@ -0,0 +1,68 @@
...
+                .authorizeRequests()
+                .antMatchers(URL_ROOT, URL_LOGIN).permitAll()
+                .anyRequest().authenticated()
+
+                .and()
+                .formLogin()
+                .loginPage(URL_LOGIN)
+                .permitAll()
+                .successHandler(authenticationSuccessHandler)
+                .failureHandler((req, resp, ex) -> resp.setStatus(SC_FORBIDDEN))
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/auth/config/WebSecurityConfig.java]: duplicate of line 56, please move it to some variable or class

File: src/main/java/by/itech/lab/supplier/controller/UserController.java [hunk 25]
```diff
@@ -1,13 +1,26 @@
 package by.itech.lab.supplier.controller;
 
+import by.itech.lab.supplier.auth.UserImpl;
+import by.itech.lab.supplier.dto.CustomerDto;
 import lombok.AllArgsConstructor;
-import org.springframework.web.bind.annotation.RequestMapping;
-import org.springframework.web.bind.annotation.RestController;
+import org.springframework.http.HttpStatus;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.web.bind.annotation.*;
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/controller/UserController.java]: wildcard imports

**Review by svareyko — APPROVED** [review]

**Pull request merged**